Learner Dashboard improvements to assist with the Account Activation process

pomegranited opened the pull request:
This change adds two optional features designed to ease learner confusion through the account activation and enrollment process.

Currently, when learners haven't yet activated their accounts, the Learner Dashboard shows both a warning about not being activated and the text below about not being enrolled in any courses. Many of our client's students don't read the first warning, get confused, and then email them about not being enrolled in the right courses.

As part of the upstream review process (see https://github.com/edx/edx-platform/pull/15464#issuecomment-362029099), we've also deprecated the top (black) banner, removed the activation message from the top banner, and permanently enabled the the sidebar activation message.

**JIRA tickets**: [OSPR-1814](https://openedx.atlassian.net/browse/OSPR-1814)

**Screenshots**:

There are three potential stages to this process, and this change improves two of them:

1. Learner has just registered, and has not yet activated their account, and is viewing the Dashboard page.
  Here is the confusing case, where the activation message in the sidebar is being missed by learners, who skip instead to the "You are not enrolled in any courses yet." message.
![show dashboard](https://user-images.githubusercontent.com/7556571/38247314-fdd12516-3783-11e8-96da-ebb7a49ac422.png)

  This change adds a Site Configuration and `FEATURE` flag, `HIDE_DASHBOARD_COURSES_UNTIL_ACTIVATED`, which hides the "My Courses" and "You are not enrolled in any courses" area.  This will ease learner confusion by making the activation message the only text shown on the page.
  Note that even if the learner was enrolled in a course prior to account activation, the system doesn't allow the user to visit the course without first activating their account.
![hide_dashboard_until_activated](https://user-images.githubusercontent.com/7556571/38247329-0ac83232-3784-11e8-8096-cd56d63c0ca5.png)

1. Learner has registered, has not yet activated their account, but has logged out or let their session expire. Learner has to login again, and/or reset their password.
  After login, the system re-sends the activation email, and shows an appropriate message, and won't let the learner proceed past the login screen until the account is activated.
  No change required here; the learner's next step is clear.
1. Learner has activated their account, is logged in, and viewing an empty dashboard with no courses in it.
  Here, we've added some optional text, which can be set in Site Configuration `EMPTY_DASHBOARD_MESSAGE`.  Our client had some specific information to convey at this point, and so configurable markup was appropriate.
![empty dashboard custom message](https://user-images.githubusercontent.com/7556571/37694913-e62de098-2d1a-11e8-89cb-3dd15b67a61b.png)

**Sandbox URL**:

https://pr15464.sandbox.opencraft.hosting (3357a25)

**Partner information**: 3rd party-hosted open edX instance

**Deployment targets**: edx.org and edge.edx.org

**Merge deadline**: ASAP

**Testing instructions**:

The `staff` user has superuser privileges on the sandbox server.
*Note* You must wait **3 min** for any Site Configuration settings changes to take effect.

1. Update the Site Configuration to set `"HIDE_DASHBOARD_COURSES_UNTIL_ACTIVATED": false`.
1. Register a new user, but don't follow the activation link yet.
   Note the "activate your account" message is visible in the Sidebar, and the top banner is not shown.
1. Update the Site Configuration to set `"HIDE_DASHBOARD_COURSES_UNTIL_ACTIVATED": true` and refresh the Dashboard.
   Note that the "My Courses" list is now hidden, and the sidebar is shifted left.
1. Visit the activation email link.
   Note that your account has been activated, and the "My Courses" list is now shown.
1. Update the Site Configuration to add an `EMPTY_DASHBOARD_MESSAGE` like the one shown in Site Configuration below.
1. Refresh the Dashboard. 
   Note that  the custom empty dashboard message is now displayed.
1. Enrol in a course, and revisit the Dashboard. 
   Note that the empty dashboard messages are no longer displayed.

**Reviewers**
- [x] @UmanShahzad 
- [ ] @marcotuts 

**Site Configuration**
```json
{
  "HIDE_DASHBOARD_COURSES_UNTIL_ACTIVATED": true,
  "EMPTY_DASHBOARD_MESSAGE": "If you believe this is an error, please contact <a href='mailto:contact@example.com'>contact@example.com</a>.<br /> If you would like to enroll in one of our free courses, you can find them in the <a href='/courses'>course catalog</a>.",
  "ACTIVATION_EMAIL_SUPPORT_LINK": "mailto:contact@example.com",
}

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=15464"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

UmanShahzad commented on common/djangoapps/student/views.py:
```diff
@@ -707,7 +707,11 @@ def dashboard(request):
+                'email': user.email,
+                'platform_name': platform_name,
+                'activation_email_support_link': activation_email_support_link,
+            }
```
Nit: Since this `dict` seems to be the same as the one above it, can we possibly factor it out?

pomegranited replied:
Good point.. addressed by https://github.com/edx/edx-platform/pull/15464/commits/c862059c609083ce196aff9a36ac6f657fc38b5d#diff-55b798ee23a7fde8d1103408afcd0f16R698

UmanShahzad commented on lms/envs/common.py:
```diff
@@ -125,6 +125,12 @@
     'COURSES_ARE_BROWSABLE': True,
 
+    # Extra message to display in the Learner Dashboard when they are not enrolled in any courses yet.
+    'EMPTY_DASHBOARD_MESSAGE': None,
```
Nit: I assume that since this is always going to be a (HTML) string (or is that incorrect?), we can default this to `''` to make it more straightforward for the user?

pomegranited replied:
Yep, that's clearer, thanks!  Addressed by https://github.com/edx/edx-platform/pull/15464/commits/c862059c609083ce196aff9a36ac6f657fc38b5d#diff-82bda515879c96ab0641bfcbe58631c3R129.

UmanShahzad commented on common/djangoapps/student/tests/test_activate_account.py:
```diff
@@ -158,7 +158,9 @@ def test_account_activation_message_disabled(self):
-            {'email': self.user.email}
+            {'email': self.user.email
+
+        }
```
Did something unintended happen here? Looks like the formatting might have been a mistake.

pomegranited replied:
Pardon me!  Fixed with de51e3d.

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-1814](https://openedx.atlassian.net/browse/OSPR-1814) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

pomegranited commented:
CC @gsong

gsong commented:
@edx/learner Can we get this nice PR on your radar?

marcotuts commented:
Hi @pomegranited, apologies for the delay. 

I was hoping to clarify some details first:

1.) Is the dashboard message still supported on open edx? I'm seeing only the dashboard sidebar area on production for edx:
![image](https://user-images.githubusercontent.com/2023680/29784675-e4407aa8-8bf2-11e7-9838-25f74bf92019.png)

2.) As you noted, no changes to the flow for logging back in. 
3.) While I can understand the need for a flag to gate viewing my courses, I don't think we should provide text overrides for this single case as a configuration, and would instead encourage us to use a theming override for this 1 case. Thoughts on that?

pomegranited commented:
Thanks for your suggestions, @marcotuts !

> 1.) Is the dashboard message still supported on open edx? I'm seeing only the dashboard sidebar area on production for edx:

Oo, it was when I submitted the PR, but looks like it's changed since.  I'll rebase and have another look -- might be Monday before I can get to that.

> 3.) While I can understand the need for a flag to gate viewing my courses, I don't think we should provide text overrides for this single case as a configuration, and would instead encourage us to use a theming override for this 1 case. Thoughts on that?

I've been trying to avoid theme customizations as much as possible, since the underlying templates shift drastically between releases, and it's more difficult to write tests for..

Would you accept a SiteConfiguration setting, like `homepage_overlay_html` instead of the feature flag?

pomegranited commented:
@marcotuts I've rebased my PR on master, and checked the following:

> 1.) Is the dashboard message still supported on open edx? I'm seeing only the dashboard sidebar area on production for edx:

edX must be using the [`DISPLAY_ACCOUNT_ACTIVATION_MESSAGE_ON_SIDEBAR` feature/site configuration flag](https://github.com/edx/edx-platform/blob/ff5c6f0c3025876f802a9375bc1173ba4db54631/common/djangoapps/student/views.py#L694-L714).  Sites without that flag set still show the dashboard message as displayed on my sandbox server.

For consistency and usability, I've also added a SiteConfiguration flag to match it, so that individual sites can override, and fixed the template change to ensure that the flag will hide the courses list regardless of where the activation message is shown.  You can verify the sidebar activation message case that edX are using by adding `"DISPLAY_ACCOUNT_ACTIVATION_MESSAGE_ON_SIDEBAR": true` to the sandbox site's [site configuration]() (`staff` user has superuser access on the sandbox).  Note that Site Configuration changes may 3 minutes to take effect.

Because doing this required a modification to the [`student.views.dashboard`](), we also have the option of optimising that method to avoid fetching enrolments for inactive learners.  I didn't do that here to keep the diff and tests small, but I'm happy to do it if you like.

> 3.) While I can understand the need for a flag to gate viewing my courses, I don't think we should provide text overrides for this single case as a configuration.

As I mentioned above, we'd rather not do a theme customization here, since it's hard to test.  But I've updated this PR to use a SiteConfiguration flag called `empty_dashboard_message`, which behaves like `homepage_overlay_html`.  I removed the `EMPTY_DASHBOARD_MESSAGE ` feature flag though, and just default to `None`, since it's not a proper on/off flag anyway.

pomegranited commented:
@marcotuts I've rebased 5da48f5 on master to resolve a conflict.

What do you think about my latest changes?  Are they acceptable?

pomegranited commented:
Rebased fe4417d on master to resolve python test error.

pomegranited commented:
Rebase 8f95f25 on master to resolve conflicts.

pomegranited commented:
Nudge @marcotuts  cc @gsong  😄

pomegranited commented:
Squashed be99869740e7599e04300d4fda941e505851838c through 07102660321e84b76a4d266a1137e481e5273a54 to make rebasing easier, and rebased on master to resolve conflicts.  Deploying a new sandbox server now to test these changes.

nedbat commented:
@marcotuts What can we do to close the loop on this?

catong commented:
@UmanShahzad @marcotuts @edx/doc If this PR merges, we should document the new site configurations.

UmanShahzad commented:
^ cc @pomegranited

pomegranited commented:
Thanks for that reminder re docs, @catong .  Once this is merged, can we just update the (now public) wiki page: [edX Feature Flags](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/34734726/edX+Feature+Flags)?

catong commented:
Hi @pomegranited It would be awesome if you could update the [edX Feature Flags](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/34734726/edX+Feature+Flags) wiki page - thank you! As well, a shorter form of the table exists in "[Index of Open edX Feature Flags](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/feature_flags/feature_flag_index.html)" in the *Installing, Configuring, and Running the Open edX Platform* guide. If you have the bandwidth, we would really appreciate your making a PR in the edx-documentation repo to mirror the wiki page update. If not, we can make a DOC JIRA ticket to follow up with updating the guide.

pomegranited commented:
jenkins run quality

pomegranited commented:
The quality failures noted here are coming from the [pavelib.quality.run_stylelint](https://build.testeng.edx.org/job/edx-platform-quality-pr/42657/console) call, and though there are a lot of errors shown there, none of them are on files I've modified for this change?  Is there not a Diff_Quality_Report anymore?

pomegranited commented:
@marcotuts  I've rebased 8c0d6b9 on master and re-deployed my sandbox.

Could I maybe get an idea of when you think there will be time to do a final review on this?

CC @nedbat

nedbat commented:
@estute what's up with the quality report on this PR?

nedbat commented:
jenkins run quality 
because of https://build.testeng.edx.org/job/edx-platform-quality-pr/42657/

marcotuts commented:
My apologies, the @ mention got lost for me in the sequence of development updates. I will connect with he team on a review this week.

pomegranited commented:
Thank you @marcotuts !

pomegranited commented:
Rebased ebc12c3 on master.

pomegranited commented:
@marcotuts CC @nedbat Any chance of getting this reviewed soon?  Thank you!

srpearce commented:
@catong Dunno if you've seen this yet, but it seems to be for learners.

pomegranited commented:
Thanks for pinging @catong , @srpearce !

@marcotuts this PR has been waiting a while, and we'd really like to get it in before Hawthorn is cut. Is there anything controversial here that I can fix to make it easier to get it through? CC @nedbat

marcotuts commented:
I will work with @chris-mike to identify the blockers here, thanks for the ping on this again @pomegranited !

marcotuts commented:
@pomegranited Hi! I had a chance to sync up with @chris-mike  about the intent of the recent activation work and have the following product suggestions / notes.

1.) We can and should move away from using the black message bar for activation purposes, and instead rely on the 2 patterns currently in use (with the waffle flag enabled) of having the sidebar activation message. My understanding is that when you return to the site and don't have an active account you are shown an alert pattern on the dashboard page in addition to the sidebar area. The implication of this is that we probably benefit from marking the existing black bar activation message area as deprecated or by removing that element entirely from edx-platform. If removal (and all that entails, including waffle flag removal) is a large ask perhaps just a note saying that this element is deprecated or marked for removal would be ok for this PR?

2.) When the HIDE_COURSES flag you are introducing is enabled, the suggestion would be to use the "well / empty" pattern we have for the case of zero enrollments but instead update this messaging to be specific to the need for activation. With this feature set, activation gates course visibility, and we can't otherwise show anything in the main content area of this page.

3.) Finally for the langauge override as I mentioned above my concern around making this configurable has to to with i18n ease, but no real product concern otherwise here. 

If you have questions I can answer let me know, otherwise I'm going to move this to engineering review. Thanks for bearing with us while we identify the owners and get all the context we needed for this. We shouldn't ever have a OSPR that drags out this log, thanks for your patience on this one!

pomegranited commented:
Thanks for your comments @marcotuts !  I'm on holiday next week, but will plan time to address your comments when I get back.

pomegranited commented:
@marcotuts @chris-mike I'm all for removing the activation banner in favour of a clearer activation pathway, but I think pushing the activation message into the sidebar makes it even harder to see.

Plus, if the inactive learner hasn't been automatically enrolled in a course, then the main content area says only "You are not enrolled in any courses yet.", which isn't very useful, when the next step is actually activation.

Would you accept a change that:
* Removes the `DISPLAY_ACCOUNT_ACTIVATION_MESSAGE_ON_SIDEBAR` feature flag and [`banner_activation_message`](https://github.com/edx/edx-platform/blob/042a854bd0191a6663c694c2ec509820c00fb913/lms/templates/dashboard.html#L89-L93) from the dashboard template.
* Adds a template comment marking the [dashboard-notifications banner](https://github.com/edx/edx-platform/blob/042a854bd0191a6663c694c2ec509820c00fb913/lms/templates/dashboard.html#L87-L119) as deprecated.
 There are other messages in there, and relocating the will take more time than I have available for this change.
* Pulls the current [inactive account activation sidebar notice](https://github.com/edx/edx-platform/blob/042a854bd0191a6663c694c2ec509820c00fb913/lms/templates/registration/account_activation_sidebar_notice.html) out of the sidebar and into the dashboard main content area, hiding/replacing the enrolled course list until the account is activated.
* The optional extra empty dashboard message added by this change is retained, but it is displayed only displays it if the account has been activated.

This resolves the confusing enrolment stages described in the PR description:
1. Learner has just registered, and has not yet activated their account, and is viewing the Dashboard page.
  The learner sees the unmissable "activate your account" message in the main content area, and is not shown the list of courses until the account is activated.
1. Learner has registered, has not yet activated their account, but has logged out or let their session expire.
  Learner has to login again, and/or reset their password.  After which, the system re-sends the activation email, shows an appropriate message, and won't let the learner proceed past the login screen until the account is activated.
  Still no change required here; the learner's next step is clear.
1. Learner has activated their account, is logged in, and viewing an empty dashboard with no courses in it.
  The default (and translated) "You are not enrolled in any courses yet." message and the "Explore courses" button are still displayed as normal.
  An optional extra (untranslatable) message can be configured per-site to sit under the empty dashboard message.

pomegranited commented:
Rebased a5a32d5 on master.

nedbat commented:
@marcotuts Any updates? (BTW: who is @chris-mike? Is that a mis-ping?)

pomegranited commented:
Thanks for your help with this @nedbat  -- it's been an open PR for an awfully long time.
@marcotuts Can you let me know when you have time to review [my proposal above](https://github.com/edx/edx-platform/pull/15464#issuecomment-365913624)?  If that sounds ok, then I'll do the work and we can hopefully, finally, get this merged!

marcotuts commented:
@pomegranited  - Thanks for the update and the details.

1.) I'm not comfortable with the change of gating access to courses until activation has happened, and would prefer the message stay in the sidebar currently. If this means we need to maintain a flag that alters this behavior in certain instances to gate access & move the massage to the main area that's fine by me, but I don't think we should hide course viewing by default on the platform. 
2.) Marking the dashboard header message area as deprecated for the account activation and removing the old flag with the old dark page level banner makes sense to me though. This assumes the new page level message (gated behind the configuration flag) would use the new alert message pattern, is that correct?

pomegranited commented:
Thanks for the feedback @marcotuts -- I'll have an update submitted by the end of next week.  Can you schedule time to do a (hopefully) final review?

pomegranited commented:
@marcotuts CC @rock345 @nedbat 

As per your https://github.com/edx/edx-platform/pull/15464#issuecomment-372475306 above, I have:
1. Marked the top banner deprecated, removed the `DISPLAY_ACCOUNT_ACTIVATION_MESSAGE_ON_SIDEBAR` feature flag, and made the sidebar activation message the default (and only) place where the activation message is displayed. a2c5fb2
1. Let the new `HIDE_DASHBOARD_COURSES_UNTIL_ACTIVATED` site configuration flag hide the My Courses list, and thereby shift the sidebar to the left side of the screen to make it more prominent. 479d4dc
1. Left the new `EMPTY_DASHBOARD_MESSAGE` site configuration flag as-is. 6c5c353

I'm redeploying the sandbox server to demonstrate this updated code, so you can use that to verify that the features work as expected.

We'd appreciate it if you could schedule this review ASAP, since this has been waiting for nearly 9 months.  Thank you!

pomegranited commented:
jenkins run bokchoy

marcotuts commented:
@pomegranited  - Thanks for the update. 

The sandbox is currently showing the sidebar message, meaning HIDE_DASHBOARD_COURSES_UNTIL_ACTIVATED is enabled. One note I should have mentioned is that in this case, I'm ok with the message being a full page message that was like what you had instead of having to exactly emulate the layout and size of the sidebar message. I'm ok with this as is if you are, but if I had any remaining suggestion it might be to make that this activation message wouldn't necessarily just render the sidebar and hide the main dashboard area. 

For now, moving to awaiting prioritization, thanks for all your help on this! 

![image](https://user-images.githubusercontent.com/2023680/37984434-415db652-31c4-11e8-8c20-a215307d8792.png)

pomegranited commented:
Oo thanks @marcotuts !  I misunderstood your previous comments, and thought I had to keep everything in the sidebar.  3357a25 shows the account activation in the main content area if `HIDE_DASHBOARD_COURSES_UNTIL_ACTIVATED` is enabled.  PR description shows updated screenshots.

I've also updated the sandbox to disable `HIDE_DASHBOARD_COURSES_UNTIL_ACTIVATED` in Site Configuration (sorry about that!).

pomegranited commented:
Rebased 2e356c4 on master to fix failing tests.

bradenmacdonald commented:
@nedbat will we still be able to get this into Hawthorn? I believe it's just waiting for a final code review and merge?

pomegranited commented:
Hi @johnmark -- is there anything you can do to help get this PR reviewed and merged?  We've had it open for so long that I'm beginning to lose hope :/

mikix commented:
@bradenmacdonald @pomegranited - Hello! I know you're trying to get this into Hawthorn. We're actively talking about how to get more Product & Engineering eyes on tickets like this. I'm specifically not making a promise here, but just giving feedback that we are aware of this ticket and trying to improve edX's process internally.

@marcotuts - is this change something that everyone would want? Like, can we just drop the feature toggle and make this improvement everywhere?

nedbat commented:
About Hawthorn: the current plan is that hawthorn.master will be created near the end of May, so there is still time.

pomegranited commented:
Nudge @marcotuts -- could you or someone on your team answer @mikix 's question above?

mduboseedx commented:
@marcotuts Did you have a chance to look at the question from @mikix?

staubina commented on lms/templates/registration/account_activation_sidebar_notice.html:
```diff
@@ -15,20 +15,7 @@ <h2 id="account-activation-title" class="account-activation-title sr">${_("Accou
-          ),
-          link_end=HTML("</a>"),
-        )}
-        </p>
```
👍 For moving this into the back end.

staubina commented:
@marcotuts From my review of your comments all your concerns have been addressed.  Is this a correct understanding?

marcotuts commented:
OK to merge from my standpoint. I wouldn't propose any changes to the flags / switches at this point.

edx-webhook commented:
@pomegranited 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/15464&entry.1484655318=https://github.com/pomegranited&entry.752974735=2017-07-05+03:44&entry.1917517419=2018-05-17+16:45&entry.2133058324=Yes) so we can improve your experience in the future.

pomegranited commented:
Oo sorry about that @michaelyoungstrom -- if I'd known it was going to be merged, I would have done one last rebase!  There's been so many :)

Have submitted https://github.com/edx/edx-platform/pull/18230, which reinstates these commits on top of the latest master.  Will ping you there once tests pass.
